### PR TITLE
Attempt to fix EBADF errors in pytest

### PR DESCRIPTION
**Pull request by keitwb**
- Upgrade Python to latest
 - Upgrade pytest and pytest-xdist
 - Don't close file descriptors for stdin/out/err when running agent as a subprocess
 - Write the config file in binary mode
 - Add decorator to catch and retry EBADF for certain functions.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -544,7 +545,7 @@ workflows:
             - helm_check
       - k8s_integration_tests:
           name: k8s_crio_integration_tests
-          k8s_version: v1.15.0
+          k8s_version: v1.16.0
```
Comment (keitwb) [inline .circleci/config.yml]: Do this in a separate PR, let's not mix things up here.

File: .circleci/config.yml [hunk 2]
```diff
@@ -368,6 +368,7 @@ jobs:
       - run:
           environment:
             K8S_VERSION: << parameters.k8s_version >>
+            WITH_CRIO: "<< parameters.with_crio >>"
```
[the remark this answers is not in the document]
Comment (jchengsfx) [reply .circleci/config.yml]: > Split to separate PR

Yep, I just needed something quick to trigger all k8s tests for now.

File: .circleci/scripts/collect_tests.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    def pytest_collection_finish(self, session):
+        for item in session.items:
+            self.collected.add(item.location[0])
+
+
+collect_tests = CollectTests()
+markers = sys.argv[1]
+directory = sys.argv[2]
+pytest.main(['--collect-only', '-m', markers, '-p', 'no:terminal', directory], plugins=[collect_tests])
```
Comment (keitwb) [inline .circleci/scripts/collect_tests.py]:
```suggestion
pytest.main(['--collect-only', '-m', markers, '-p', 'no:terminal', directory], plugins=[CollectTests()])
```

File: tests/helpers/util.py [hunk 4]
```diff
@@ -156,11 +157,15 @@ def get_unique_localhost():
...
-    )
+    try:
+        proc = subprocess.Popen(
+            [str(c) for c in command], env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, **kwargs
+        )
+    except OSError as e:
+        sys.stderr.write(f"{e}\n")
 
     get_output = pull_from_reader_in_background(proc.stdout)
+    assert proc.poll() is None, f"agent is not running:\n{get_output}"
```
Comment (keitwb) [inline tests/helpers/util.py]: `get_output` is a function, not a string value.

Also this is super racy.  If the agent dies immediately after this check it will still go on happily.

Comment (keitwb) [reply tests/helpers/util.py]: Also `proc` will be None if the OSError happens since it will throw the exception before `proc` gets assigned.
